llvm/test/Transforms/InstCombine/abs-intrinsic.ll

```diff
@@ -5,8 +5,57 @@ declare i8 @llvm.abs.i8(i8, i1)
 declare i32 @llvm.abs.i32(i32, i1)
 declare <4 x i32> @llvm.abs.v4i32(<4 x i32>, i1)
 declare <3 x i82> @llvm.abs.v3i82(<3 x i82>, i1)
+declare <2 x i8> @llvm.abs.v2i8(<2 x i8>, i1)
 declare void @llvm.assume(i1)
 
+define i8 @test_abs_abs_a_mul_b_i8(i8 %a, i8 %b) {
+; CHECK-LABEL: @test_abs_abs_a_mul_b_i8(
+; CHECK-NEXT:    [[TMP1:%.*]] = mul i8 [[B:%.*]], [[A:%.*]]
+; CHECK-NEXT:    [[ABS2:%.*]] = call i8 @llvm.abs.i8(i8 [[TMP1]], i1 true)
+; CHECK-NEXT:    ret i8 [[ABS2]]
+;
+  %abs1 = call i8 @llvm.abs.i8(i8 %a, i1 true)
+  %mul = mul i8 %abs1, %b
+  %abs2 = call i8 @llvm.abs.i8(i8 %mul, i1 true)
+  ret i8 %abs2
+}
+
+define i8 @test_abs_a_mul_abs_b_i8(i8 %a, i8 %b) {
+; CHECK-LABEL: @test_abs_a_mul_abs_b_i8(
+; CHECK-NEXT:    [[TMP1:%.*]] = mul i8 [[A:%.*]], [[B:%.*]]
+; CHECK-NEXT:    [[ABS2:%.*]] = call i8 @llvm.abs.i8(i8 [[TMP1]], i1 true)
+; CHECK-NEXT:    ret i8 [[ABS2]]
+;
+  %abs1 = call i8 @llvm.abs.i8(i8 %b, i1 true)
+  %mul = mul i8 %abs1, %a
```

Comment:
@nikic  what is `thwart complexity-based canonicalization` ?
and I think we should add oneUse check to the internal `abs`, right?

Reply:
> @nikic what is `thwart complexity-based canonicalization` ?

`git grep "thwart complexity-based canonicalization"` :) It's used to prevent InstCombine from flipping the opernd order

> and I think we should add oneUse check to the internal `abs`, right?

You don't need OneUse on the abs, but you do need it on the `mul`.

Reply:
Okay, but I need a little clarification.
we use OneUse to make sure that the instruction can be removed safely, so we need it on the `mul`, but what if we need the value of `abs(x)` later, I know it'll be folded later for example `x + abs(y)`.

Reply:
In that case we would not reduce the number of instructions, but we would still "simplify" the mul by reducing the dependency chain (it now uses y directly instead of abs(y)). Unless there is some specific reason to the contrary, we allow such transforms.